lib/src/authentication/scheme.dart

```diff
@@ -0,0 +1,77 @@
+import 'credential.dart';
+
+// TODO(themisir): Add authentication scopes. It'll be used to check whether or
+// not the request should be authenticated. For example official pub.dev server
+// scheme could define scopes as [AuthenticationScope.publish] which means it
+// only requires authentication for publish command.
+
+/// Abstract interface for authentication scheme. It's used for validating
+/// request URLs and contains [credential] which will be used for
+/// authentication.
+abstract class AuthenticationScheme {
+  /// Authentication credential which used to resolve `Authorization` header
+  /// value.
+  Credential get credential;
```

Comment:
I've been digging into this a bit the past few days... My two cents is that having abstractions:
 (A) `AuthenticationScheme`, and,
 (B) `Credential`,
is perhaps a bit much.

I don't have a solution. But I'm tiny bit concern we're making this a lot more complicated than it needs to be.

I'll play with this a bit more later.